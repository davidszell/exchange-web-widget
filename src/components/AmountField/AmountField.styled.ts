import styled from 'styled-components';

export const Input = styled.input`
  flex: 1;
  border: none;
  text-align: right;
  font-size: 1.3em;
  padding: 10px;
  border-radius: 5px;
  -moz-appearance: textfield;

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
`;
