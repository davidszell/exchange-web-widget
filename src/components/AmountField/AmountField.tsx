import React, { useEffect, useState } from 'react';
import { formatFloatFloor } from '../../helpers';
import { Input } from './AmountField.styled';

type AmountFieldPropType = {
  amount: number | null,
  handleAmountChange: (newAmount: number | null) => void
};

const AmountField = ({ amount, handleAmountChange }: AmountFieldPropType): JSX.Element => {
  const [displayValue, setDisplayValue] = useState<number | ''>('');

  useEffect(() => {
    setDisplayValue(amount === null ? '' : amount);
  }, [amount]);

  const handleOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { valueAsNumber, value: valueFromInput } = event.target;
    if (valueFromInput === '' || Number.isNaN(valueAsNumber) || valueAsNumber <= 0) {
      handleAmountChange(null);
      return;
    }

    handleAmountChange(formatFloatFloor(valueAsNumber));
  };

  return (
    <Input
      data-testid="amount"
      type="number"
      value={displayValue}
      step="0.01"
      onChange={handleOnChange}
      placeholder="0"
    />
  );
};

export default AmountField;
