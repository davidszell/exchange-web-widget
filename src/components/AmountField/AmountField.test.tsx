import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import AmountField from './AmountField';
import { fireEvent, render, screen } from '@testing-library/react';

describe('AmountField', () => {
  it('matches snapshot', () => {
    const tree = renderer
      .create(<AmountField amount={10} handleAmountChange={() => { }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('renders correctly when defualt amount is set to number', () => {
    render(<AmountField amount={10} handleAmountChange={() => { }} />);
    expect(screen.getByTestId('amount')).toHaveDisplayValue('10');
  })

  it('renders correctly when defualt amount is set to null', () => {
    render(<AmountField amount={null} handleAmountChange={() => { }} />);
    expect(screen.getByTestId('amount')).toHaveDisplayValue('');
  })

  it('calls handleAmountChange when amount changes', () => {
    const handleAmountMock = jest.fn();
    render(<AmountField amount={10} handleAmountChange={handleAmountMock} />);
    fireEvent.change(screen.getByTestId('amount'), {target: {value: '20'}});
    expect(handleAmountMock).toBeCalled();
  })

  it('calls handleAmountChange with correct value', () => {
    const handleAmountMock = jest.fn();
    render(<AmountField amount={10} handleAmountChange={handleAmountMock} />);
    fireEvent.change(screen.getByTestId('amount'), {target: {value: '20'}});
    expect(handleAmountMock).toBeCalledWith(20);
  })

  it('calls handleAmountChange with null on non-number', () => {
    const handleAmountMock = jest.fn();
    render(<AmountField amount={10} handleAmountChange={handleAmountMock} />);
    fireEvent.change(screen.getByTestId('amount'), {target: {value: 'foo'}});
    expect(handleAmountMock).toBeCalledWith(null);
  })

  it('calls handleAmountChange with null on 0', () => {
    const handleAmountMock = jest.fn();
    render(<AmountField amount={10} handleAmountChange={handleAmountMock} />);
    fireEvent.change(screen.getByTestId('amount'), {target: {value: '0'}});
    expect(handleAmountMock).toBeCalledWith(null);
  })

  it('calls handleAmountChange with null on negative number', () => {
    const handleAmountMock = jest.fn();
    render(<AmountField amount={10} handleAmountChange={handleAmountMock} />);
    fireEvent.change(screen.getByTestId('amount'), {target: {value: '-1'}});
    expect(handleAmountMock).toBeCalledWith(null);
  })

  it('calls handleAmountChange with correct float formatting', () => {
    const handleAmountMock = jest.fn();
    render(<AmountField amount={10} handleAmountChange={handleAmountMock} />);
    fireEvent.change(screen.getByTestId('amount'), {target: {value: '1.234'}});
    expect(handleAmountMock).toBeCalledWith(1.23);
  })
})