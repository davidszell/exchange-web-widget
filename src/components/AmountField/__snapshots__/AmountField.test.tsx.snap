// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountField matches snapshot 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  border: none;
  text-align: right;
  font-size: 1.3em;
  padding: 10px;
  border-radius: 5px;
  -moz-appearance: textfield;
}

.c0::-webkit-outer-spin-button,
.c0::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

<input
  className="c0"
  data-testid="amount"
  onChange={[Function]}
  placeholder="0"
  step="0.01"
  type="number"
  value=""
/>
`;
